src/libAtomVM/atom_table.c

```diff
@@ -482,7 +540,14 @@ int atom_table_ensure_atoms(
                 }
             }
 
-            unsigned long hash = sdbm_hash(current_atom, atom_string_len(current_atom));
+            if (is_long_format) {
+                uint8_t *atom_copy = malloc(atom_len + 1);
```

Comment:
There is no check that allocation succeeds here.

Could we somehow avoid the copy as we do if !is_long_format? For example we could alter the structure of a node (struct HNode).

Reply:
`HNode` is meant to be a fixed size struct so I cannot save it there.
My opinion here is to add a PackBEAM step that "optimizes" atoms by reverting them to the fixed length encoding, so no additional memory will be used.
On the long term I would like to have a shared atom table for each .avm file that uses the right format and that saves space from a number of duplicated atoms, but this is another (OT) story.

TL;DR: let's keep this as  is just for .beam files that didn't receive any further processing.

Reply:
> There is no check that allocation succeeds here.

Fixed.

Reply:
I'm a little bit confused about `HNode` is a fixed size struct.

Couldn't we do:

```
struct HNode
{
    struct HNode *next;
    uint8_t len;
    const char* key;
    long index;
};
```

This would of course mean more RAM for the default case, but we could eventually save some without altering avm format by sharing pointer to key with common prefix search or substrings.

Also we could also limit the number of atoms to 2^24 and pack index and len on the same uint32_t.

Reply:
Wait, ok, maybe we've been talking about 2 different things.
It sounds like a clever change, but I think it might turn into a very big change way beyond adding OTP-28 support.